src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -28,45 +140,79 @@ public Future<TokenUsage> getTokenStats(ProxyContext context) {
         return traceContext.getStats(context);
     }
 
-    public void endSpan(ProxyContext context) {
+    public Future<Void> endSpan(ProxyContext context) {
         ApiKeyData apiKeyData = context.getApiKeyData();
         if (apiKeyData.getPerRequestKey() == null) {
             traceIdToContext.remove(context.getTraceId());
         } else {
             TraceContext traceContext = traceIdToContext.get(context.getTraceId());
             if (traceContext != null) {
-                traceContext.endSpan(context);
+                return traceContext.endSpan(context);
             }
         }
+        return Future.succeededFuture();
     }
 
     @VisibleForTesting
     Map<String, TraceContext> getTraces() {
         return traceIdToContext;
     }
 
-    private static class TraceContext {
+    @VisibleForTesting
+    Map<String, TrackerData> getRegisteredTrackers() {
+        return registeredTrackers;
+    }
+
+    private class TraceContext {
         Map<String, TokenStats> spans = new HashMap<>();
 
-        synchronized void addSpan(ProxyContext context) {
+        synchronized Future<Void> addSpan(ProxyContext context) {
             String spanId = context.getSpanId();
             String parentSpanId = context.getParentSpanId();
-            TokenStats tokenStats = new TokenStats(new TokenUsage(), Promise.promise(), new ArrayList<>());
+            TokenStats tokenStats = new TokenStats(new TokenUsage(), new HashMap<>());
             spans.put(spanId, tokenStats);
             if (parentSpanId != null) {
-                TokenStats parent = spans.get(parentSpanId);
-                if (parent != null) {
-                    parent.children.add(tokenStats.promise.future());
-                }
+                StartSpanEvent event = new StartSpanEvent(trackerId, context.getTraceId(), parentSpanId, context.getSpanId());
+                return publish(Event.START_SPAN, event);
```

Comment:
With all this future code it is hard to tell if this is possible:
1) Flow: 1 -> 2 -> 3. -> means edge, so 1 is parent span for 2.
2) Real publication order: 1 -> 3 -> 2.

Reply:
I can't imagine the case.
Core aways startSpan parent span before child span. So events will come in the order they were sent to the topic.
The order of End span events doesn't matter.